### PR TITLE
[2.5] F3 Streaming Code Rewrite

### PR DESCRIPTION
**Pull request by nvidianz**
Fixes FLARE-2241

1. Refactored F3 streaming code so it's more object oriented, easier to debug.
2. Fixed a few corner cases found during testing with a testing driver that injects duplicate packets.

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Quick tests passed locally by running `./runtest.sh`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated.

### CODE REVIEW

File: nvflare/fuel/f3/streaming/byte_receiver.py [hunk 1]
```diff
@@ -41,115 +41,226 @@
...
 
 class RxTask:
     """Receiving task for ByteStream"""
 
-    def __init__(self, sid: int, origin: str):
+    rx_task_map = {}
+    map_lock = threading.Lock()
+    timeout = CommConfigurator().get_streaming_read_timeout(READ_TIMEOUT)
+    ack_interval = CommConfigurator().get_streaming_ack_interval(ACK_INTERVAL)
+    max_out_seq = CommConfigurator().get_streaming_max_out_seq_chunks(MAX_OUT_SEQ_CHUNKS)
```
Comment (yanchengnv) [inline nvflare/fuel/f3/streaming/byte_receiver.py:58]: This requires that the CommConfigurator must be fully working at the time that the RxTask class is initialized. Not sure this can be met.

Comment (nvidianz) [reply nvflare/fuel/f3/streaming/byte_receiver.py]: Moved to constructor.

File: nvflare/fuel/f3/streaming/byte_receiver.py [hunk 2]
```diff
@@ -41,115 +41,244 @@
...
+    def __init__(self, task: RxTask):
+        super().__init__(task.size, task.headers)
+        self.task = task
+
+    def read(self, size: int) -> bytes:
+        if self.closed:
+            raise StreamError("Read from closed stream")
+
+        return self.task.read(size)
+
```
Comment (YuanTingHsieh) [inline nvflare/fuel/f3/streaming/byte_receiver.py:281]: what's the diff. RxTask VS RxStream? should we just merge RxTask into RxStream?

File: nvflare/fuel/f3/streaming/byte_streamer.py [hunk 3]
```diff
@@ -71,17 +82,162 @@ def __init__(
...
+        self.stream_future = StreamFuture(self.sid)
+        self.stream_future.set_size(stream.get_size())
+
+        self.window_size = CommConfigurator().get_streaming_window_size(STREAM_WINDOW_SIZE)
+        self.ack_wait = CommConfigurator().get_streaming_ack_wait(STREAM_ACK_WAIT)
 
     def __str__(self):
         return f"Tx[SID:{self.sid} to {self.target} for {self.channel}/{self.topic}]"
 
+    def send_loop(self):
```
Comment (YuanTingHsieh) [inline nvflare/fuel/f3/streaming/byte_streamer.py:96]: For this underlying change, is it low-risk? If QA can't test 2.5.1 thoroughly, should we move this PR to main?

File: nvflare/fuel/f3/streaming/byte_receiver.py (hunk 2 above)
Comment (nvidianz) [reply nvflare/fuel/f3/streaming/byte_receiver.py:281]: RxStream is an implementation of Stream interface. RxTask is used to hike the details of a streaming task. Better to keep them separate.

File: nvflare/fuel/f3/streaming/byte_streamer.py (hunk 3 above)
Comment (nvidianz) [reply nvflare/fuel/f3/streaming/byte_streamer.py:96]: The risk is pretty low but it's a good idea to merge the change into main. All the bugs we discovered are already fixed in 2.5.

Comment (nvidianz) [reply nvflare/fuel/f3/streaming/byte_streamer.py:96]: Created https://github.com/NVIDIA/NVFlare/pull/2960 for main branch.

File: nvflare/fuel/f3/streaming/byte_streamer.py [hunk 4]
```diff
@@ -51,10 +51,21 @@
 
 class TxTask:
     def __init__(
-        self, channel: str, topic: str, target: str, headers: dict, stream: Stream, secure: bool, optional: bool
+        self,
```
Comment (yanchengnv) [inline nvflare/fuel/f3/streaming/byte_streamer.py:54]: Not sure about this seemingly massive changes. Why are these necessary?

**Review by yanchengnv — CHANGES_REQUESTED** [review]
Some further improvements suggested. See my comments.

File: nvflare/fuel/f3/streaming/byte_receiver.py [hunk 5]
```diff
@@ -182,148 +308,53 @@ def register_callback(self, channel: str, topic: str, stream_cb: Callable, *args
...
         sid = message.get_header(StreamHeaderKey.STREAM_ID)
         origin = message.get_header(MessageHeaderKey.ORIGIN)
         seq = message.get_header(StreamHeaderKey.SEQUENCE)
         error = message.get_header(StreamHeaderKey.ERROR_MSG, None)
 
-        payload = message.payload
-
-        with self.map_lock:
-            task = self.rx_task_map.get(sid, None)
+        with RxTask.map_lock:
```
[the remark this answers is not in the document]
Comment (nvidianz) [reply nvflare/fuel/f3/streaming/byte_receiver.py]: Refactored as suggested.

File: nvflare/fuel/f3/streaming/byte_streamer.py (hunk 4 above)
Comment (nvidianz) [reply nvflare/fuel/f3/streaming/byte_streamer.py]: Removed the changes from 2.5.

**Review by yanchengnv — APPROVED** [review]
LGTM
